### PR TITLE
added sfixes for reducers

### PR DESCRIPTION
**Pull request by makart19**

### CODE REVIEW

File: cpp/daal/src/sycl/reducer.cpp [hunk 1]
```diff
@@ -63,23 +64,37 @@ services::Status Reducer::buildProgram(ClKernelFactoryIface & kernelFactory, con
...
-    DAAL_CHECK_STATUS_RETURN_VOID_IF_FAIL(status);
+    DAAL_CHECK_STATUS_VAR(status);
+
+    // no need to check overflow for nVectors * vectorSize due to we already have buffer vectors of such size
+    if (vectors.type() == TypeIds::id<float>())
+    {
+        DAAL_ASSERT_UNIVERSAL_BUFFER(vectors, float, nVectors * vectorSize);
+        DAAL_ASSERT_UNIVERSAL_BUFFER(reduceRes, float, nVectors);
+    }
+    else
```
Comment (amgrigoriev) [inline cpp/daal/src/sycl/reducer.cpp:81]: Maybe to check that vector is double and report error if both not float and not double?

File: cpp/daal/src/sycl/reducer.cpp [hunk 2]
```diff
@@ -88,60 +103,92 @@ void Reducer::singlepass(ExecutionContextIface & context, ClKernelFactoryIface &
...
+    if (vectors.type() == TypeIds::id<float>())
+    {
+        DAAL_ASSERT_UNIVERSAL_BUFFER(vectors, float, nVectors * vectorSize);
+        DAAL_ASSERT_UNIVERSAL_BUFFER(stepResult.reduceRes, float, nVectors * numDivisionsByRow);
+    }
+    else
+    {
+        DAAL_ASSERT_UNIVERSAL_BUFFER(vectors, double, nVectors * vectorSize);
+        DAAL_ASSERT_UNIVERSAL_BUFFER(stepResult.reduceRes, double, nVectors * numDivisionsByRow);
+    }
```
Comment (amgrigoriev) [inline cpp/daal/src/sycl/reducer.cpp:127]: The same

File: cpp/daal/src/sycl/reducer.cpp [hunk 3]
```diff
@@ -88,60 +103,92 @@ void Reducer::singlepass(ExecutionContextIface & context, ClKernelFactoryIface &
...
+    if (result.reduceRes.type() == TypeIds::id<float>())
+    {
+        DAAL_ASSERT_UNIVERSAL_BUFFER(stepResult.reduceRes, float, nVectors * vectorSize);
+        DAAL_ASSERT_UNIVERSAL_BUFFER(result.reduceRes, float, nVectors);
+    }
+    else
+    {
+        DAAL_ASSERT_UNIVERSAL_BUFFER(stepResult.reduceRes, double, nVectors * vectorSize);
+        DAAL_ASSERT_UNIVERSAL_BUFFER(result.reduceRes, double, nVectors);
+    }
```
Comment (amgrigoriev) [inline cpp/daal/src/sycl/reducer.cpp:169]: the same

File: cpp/daal/src/sycl/sum_reducer.cpp [hunk 4]
```diff
@@ -46,23 +47,40 @@ services::Status buildProgram(ClKernelFactoryIface & kernelFactory, const TypeId
...
+        DAAL_ASSERT_UNIVERSAL_BUFFER(vectors, float, nVectors * vectorSize);
+        DAAL_ASSERT_UNIVERSAL_BUFFER(result.sum, float, nVectors);
+        DAAL_ASSERT_UNIVERSAL_BUFFER(result.sumOfSquares, float, nVectors);
+    }
+    else
+    {
+        DAAL_ASSERT_UNIVERSAL_BUFFER(vectors, double, nVectors * vectorSize);
+        DAAL_ASSERT_UNIVERSAL_BUFFER(result.sum, double, nVectors);
+        DAAL_ASSERT_UNIVERSAL_BUFFER(result.sumOfSquares, double, nVectors);
+    }
```
Comment (amgrigoriev) [inline cpp/daal/src/sycl/sum_reducer.cpp:71]: the same

File: cpp/daal/src/sycl/sum_reducer.cpp [hunk 5]
```diff
@@ -101,25 +139,46 @@ void runStepColmajor(ExecutionContextIface & context, ClKernelFactoryIface & ker
...
+        DAAL_ASSERT_UNIVERSAL_BUFFER(result.sum, float, nVectors);
+        DAAL_ASSERT_UNIVERSAL_BUFFER(result.sumOfSquares, float, nVectors);
+    }
+    else
+    {
+        DAAL_ASSERT_UNIVERSAL_BUFFER(stepResult.sum, double, nVectors * vectorSize);
+        DAAL_ASSERT_UNIVERSAL_BUFFER(stepResult.sumOfSquares, double, nVectors * vectorSize);
+        DAAL_ASSERT_UNIVERSAL_BUFFER(result.sum, double, nVectors);
+        DAAL_ASSERT_UNIVERSAL_BUFFER(result.sumOfSquares, double, nVectors);
+    }
```
Comment (amgrigoriev) [inline cpp/daal/src/sycl/sum_reducer.cpp:167]: the same

File: cpp/daal/src/sycl/reducer.cpp [hunk 6]
```diff
@@ -161,6 +209,8 @@ Reducer::Result Reducer::reduce(const BinaryOp op, Layout vectorsLayout, const U
     Result result(context, resReduce, nVectors, vectors.type(), status);
     DAAL_CHECK_STATUS_RETURN_IF_FAIL(status, Reducer::Result());
 
+    DAAL_ASSERT(vectors.type() == TypeIds::id<float>() || vectors.type() == TypeIds::id<float>());
```
Comment (amgrigoriev) [inline cpp/daal/src/sycl/reducer.cpp:212]: double?

Comment (makart19) [reply cpp/daal/src/sycl/reducer.cpp]: fixed

**Review by amgrigoriev — APPROVED** [review]

**Comment by makart19** [comment]
@Mergifyio backport rls/2021-gold-mnt

**Pull request merged**